File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-1.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        minimum = nums[0]
        l, r = 0, len(nums) - 1

        while l <= r:
            # Array already sorted
            if nums[l] < nums[r]:
                minimum = min(minimum, nums[l])
                break
            
            m = (l + r) // 2
            minimum = min(minimum, nums[m])
            if nums[m] >= nums[l]:
                l = m + 1
            else:
                r = m - 1
        return minimum

File: Data Structures & Algorithms/eating-bananas/submission-1.cpp
// Brute Force
// T: O(n * m), S: O(1)
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        // Find the largest pile.
        int maxPile = 0;
        for (int pile : piles) {
            maxPile = max(maxPile, pile);
        }

        // Try every possible eating speed.
        for (int k = 1; k <= maxPile; k++) {
            int totalHours = 0;

            for (int pile : piles) {
                // Hours needed for this pile: ceil(pile / k)
                totalHours += (pile + k - 1) / k; // Somehow rounding up. For example for 11 / 3 we want 4

                // No need to continue if we already exceeded h.
                if (totalHours > h) {
                    break;
                }
            }

            // First valid speed is the minimum answer.
            if (totalHours <= h) {
                return k;
            }
        }

        return -1; // Should never happen for valid inputs.
    }
};
